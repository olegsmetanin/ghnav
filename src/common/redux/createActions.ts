export const createActions = (entityName: string) => {
  const actionTypes = {
    FAILURE: `${entityName}/FAILURE`,
    LOAD: `${entityName}/LOAD`,
    LOAD_SUCCESS: `${entityName}/LOAD_SUCCESS`
  }

  function failure(error) {
    return {
      type: actionTypes.FAILURE,
      error
    }
  }

  function load(query: any) {
    return {
      type: actionTypes.LOAD,
      payload: query
    }
  }

  function loadSuccess(valueNquery: any) {
    return {
      type: actionTypes.LOAD_SUCCESS,
      payload: valueNquery
    }
  }

  return {
    actionTypes,
    failure,
    load,
    loadSuccess
  }
}
