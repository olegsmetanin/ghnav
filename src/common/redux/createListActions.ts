export const createListActions = (entityName: string) => {
  const actionTypes = {
    FAILURE: `${entityName}_LIST/FAILURE`,
    LOAD: `${entityName}_LIST/LOAD`,
    LOAD_SUCCESS: `${entityName}_LIST/LOAD_SUCCESS`,
    LOADING: `${entityName}_LIST/LOADING`
  }

  function failure(error) {
    return {
      type: actionTypes.FAILURE,
      error
    }
  }

  function load(query: any) {
    return {
      type: actionTypes.LOAD,
      payload: query
    }
  }

  function loadSuccess(valueNquery: any) {
    return {
      type: actionTypes.LOAD_SUCCESS,
      payload: valueNquery
    }
  }

  return {
    actionTypes,
    failure,
    load,
    loadSuccess
  }
}
