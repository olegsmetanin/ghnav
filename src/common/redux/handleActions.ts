export function handleActions<T>(handlers: { [key: string]: (state: T, action: any) => T }, initialState: T) {
  return (state: T = initialState, action: any) =>
    action && handlers[action.type] ? handlers[action.type](state, action) : state
}
