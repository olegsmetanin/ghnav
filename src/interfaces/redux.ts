// Flux Standard Action
export interface IFSA<P, M> {
  type: string
  payload?: P
  error: IError
  meta: M
}

export interface IProcessState {
  isLoading?: boolean

  isLoadingMore?: boolean

  isSaving?: boolean

  isDeleting?: boolean
}

export interface IError {
  [zone: string]: Error[]
}
