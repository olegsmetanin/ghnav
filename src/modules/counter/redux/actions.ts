export const actionTypes = {
  INCREMENT: 'COUNTER/INCREMENT',
  INCREMENT_SUCCESS: 'COUNTER/INCREMENT_SUCCESS',
  INCREMENT_FAILURE: 'COUNTER/INCREMENT_FAILURE'
}

export function increment(value: number) {
  return {
    type: actionTypes.INCREMENT,
    payload: value
  }
}

export function incrementSuccess(value: number) {
  return {
    type: actionTypes.INCREMENT_SUCCESS,
    payload: value
  }
}

export function incrementFailure(error) {
  return {
    type: actionTypes.INCREMENT_FAILURE,
    error
  }
}
