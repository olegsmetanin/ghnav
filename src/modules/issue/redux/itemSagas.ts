/* global fetch */

import 'isomorphic-unfetch'

import { actionTypes, failure, loadSuccess } from './itemActions'
import { put, takeLatest } from 'redux-saga/effects'

// import jsontab from 'modules/issue/containers/__tests__/issue.stab.json'

export function* loadSaga(action) {
  try {
    const query = action.payload
    const url = `https://api.github.com/repos/${query.owner}/${query.repo}/issues/${query.num}`
    const response = yield fetch(url)
    if (response.status < 200 || response.status > 300) {
      throw new Error(response.statusText)
    }
    const value = yield response.json()
    yield put(loadSuccess({ query, value }))
  } catch (err) {
    yield put(failure(err.message))
  }
}

export const itemSagas = [takeLatest(actionTypes.LOAD, loadSaga)]
