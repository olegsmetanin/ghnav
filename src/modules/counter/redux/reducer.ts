import { actionTypes } from './actions'

export const initialState = {
  value: 0
}

export function reducer(state = initialState, action) {
  switch (action.type) {
    case actionTypes.INCREMENT_SUCCESS:
      return {
        ...state,
        ...{
          value: state.value + action.payload
        }
      }

    default:
      return state
  }
}
