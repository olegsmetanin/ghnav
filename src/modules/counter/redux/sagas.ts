import { actionTypes, incrementFailure, incrementSuccess } from './actions'
import { put, takeLatest } from 'redux-saga/effects'

export function* incrementSaga(action) {
  try {
    const value = action.payload

    if (typeof value === 'number') {
      yield put(incrementSuccess(value))
    } else {
      throw new Error('increment value must be a number')
    }
  } catch (err) {
    yield put(incrementFailure(err.message))
  }
}

export const sagas = [takeLatest(actionTypes.INCREMENT, incrementSaga)]
