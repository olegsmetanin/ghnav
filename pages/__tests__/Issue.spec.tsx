/* eslint-env jest */
import Issue from '../issue'
import { Provider } from 'react-redux'
import React from 'react'
import { all } from 'redux-saga/effects'
import configureMockStore from 'redux-mock-store'
import createSagaMiddleware from 'redux-saga'
import fakeAPIResponse from 'modules/issue/containers/__tests__/issue.stab.json'
import fetchMock from 'fetch-mock'
import { itemSagas as issueSagas } from 'modules/issue/redux/itemSagas'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
// import { rootSaga } from 'modules/main/rootSaga'

const sagaMiddleware = createSagaMiddleware()
const mockStore = configureMockStore([sagaMiddleware])

describe('Issue', () => {
  afterEach(fetchMock.restore)

  it('renders correct', async () => {
    const store = mockStore({ issue: { value: null } })
    const props = await Issue.getInitialProps({
      ctx: {
        store,
        asPath: '/issue?owner=zeit&repo=next.js&num=5638'
      }
    })
    const component = renderer.create(
      <Provider store={store}>
        <Issue {...props} />
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('works correct with store', async () => {
    const store = mockStore({ issue: { value: null } })

    function* rootSaga() {
      yield all([...issueSagas])
    }

    store.runSagaTask = () => {
      store.sagaTask = sagaMiddleware.run(rootSaga)
    }

    store.runSagaTask()

    fetchMock.get('https://api.github.com/repos/zeit/next.js/issues/5638', {
      status: 200,
      body: fakeAPIResponse
    })

    const props = await Issue.getInitialProps({
      ctx: {
        store,
        asPath: '/issue?owner=zeit&repo=next.js&num=5638'
      }
    })

    await new Promise(resolve => setImmediate(resolve))

    expect(store.getActions().map(el => el.type)).toEqual(['ISSUE/LOAD', 'ISSUE/LOAD_SUCCESS'])
  })

  it('works with errors', async () => {
    const store = mockStore({ issue: { value: null } })

    function* rootSaga() {
      yield all([...issueSagas])
    }

    store.runSagaTask = () => {
      store.sagaTask = sagaMiddleware.run(rootSaga)
    }

    store.runSagaTask()

    fetchMock.get('https://api.github.com/repos/zeit/next.js/issues/5638', {
      status: 404,
      body: { error: 'no data' }
    })

    const props = await Issue.getInitialProps({
      ctx: {
        store,
        asPath: '/issue?owner=zeit&repo=next.js&num=5638'
      }
    })

    await new Promise(resolve => setImmediate(resolve))

    expect(store.getActions().map(el => el.type)).toEqual(['ISSUE/LOAD', 'ISSUE/FAILURE'])
  })
})
