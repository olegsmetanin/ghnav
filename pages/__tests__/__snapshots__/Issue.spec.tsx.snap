// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issue renders correct 1`] = `
Array [
  <div
    className="PrimarySearchAppBar-root-2"
  >
    <header
      className="MuiPaper-root-21 MuiPaper-elevation4-27 MuiAppBar-root-12 MuiAppBar-positionFixed-13 MuiAppBar-colorPrimary-19 mui-fixed"
    >
      <div
        className="MuiToolbar-root-48 MuiToolbar-dense-51 MuiToolbar-gutters-49"
      >
        <button
          aria-label="Open drawer"
          className="MuiButtonBase-root-58 MuiIconButton-root-52 MuiIconButton-colorInherit-53 PrimarySearchAppBar-menuButton-4"
          disabled={false}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiIconButton-label-57"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-61"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-128"
          />
        </button>
        <a
          className="IssuePage-backLink-1"
          href="/?owner=zeit&repo=next.js"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-58 MuiIconButton-root-52 MuiIconButton-colorInherit-53"
            disabled={false}
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-57"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-61"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-128"
            />
          </button>
        </a>
        <h5
          className="MuiTypography-root-70 MuiTypography-h5-86 MuiTypography-colorInherit-99 MuiTypography-noWrap-96"
        >
          Github 
          zeit
          /
          next.js
           #
          5638
        </h5>
        <div
          className="PrimarySearchAppBar-grow-3"
        />
        <div
          className="PrimarySearchAppBar-sectionDesktop-10"
        >
          <button
            className="MuiButtonBase-root-58 MuiIconButton-root-52 MuiIconButton-colorInherit-53"
            disabled={false}
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-57"
            >
              <span
                className="MuiBadge-root-105"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root-61"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
                  />
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                </svg>
                <span
                  className="MuiBadge-badge-106 MuiBadge-colorSecondary-108"
                >
                  4
                </span>
              </span>
            </span>
            <span
              className="MuiTouchRipple-root-128"
            />
          </button>
          <button
            className="MuiButtonBase-root-58 MuiIconButton-root-52 MuiIconButton-colorInherit-53"
            disabled={false}
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-57"
            >
              <span
                className="MuiBadge-root-105"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root-61"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 22c1.1 0 2-.9 2-2h-4c0 1.1.89 2 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z"
                  />
                </svg>
                <span
                  className="MuiBadge-badge-106 MuiBadge-colorSecondary-108"
                >
                  17
                </span>
              </span>
            </span>
            <span
              className="MuiTouchRipple-root-128"
            />
          </button>
          <button
            aria-haspopup="true"
            aria-owns={null}
            className="MuiButtonBase-root-58 MuiIconButton-root-52 MuiIconButton-colorInherit-53"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-57"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-61"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-128"
            />
          </button>
        </div>
        <div
          className="PrimarySearchAppBar-sectionMobile-11"
        >
          <button
            aria-haspopup="true"
            className="MuiButtonBase-root-58 MuiIconButton-root-52 MuiIconButton-colorInherit-53"
            data-test-id="mobile-menu-icon"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-57"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-61"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-128"
            />
          </button>
        </div>
      </div>
    </header>
  </div>,
  <div
    className="BaseIssueOverview-root-115"
  >
    <div
      style={
        Object {
          "display": "flex",
          "marginTop": 100,
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "margin": "auto",
          }
        }
      >
        <div
          className="MuiCircularProgress-root-118 MuiCircularProgress-colorPrimary-121 MuiCircularProgress-indeterminate-120"
          role="progressbar"
          style={
            Object {
              "height": 40,
              "width": 40,
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg-123"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle-124 MuiCircularProgress-circleIndeterminate-126"
              cx={44}
              cy={44}
              fill="none"
              r={20.2}
              strokeWidth={3.6}
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
]
`;
